tasks/main.yml

```diff
@@ -63,9 +63,38 @@
     name: dell-system-update
     state: latest
 
+- name: Set upgrade_categories when dell_dsu_update_all_firmware is True
+  set_fact:
+    upgrade_categories: all
+  when:
+    - upgrade_categories is not defined
+    - (dell_dsu_update_all_firmware is defined) and (dell_dsu_update_all_firmware)
+
+- name: Set categories_list
+  set_fact:
+    categories_list: "{{ upgrade_categories.split(',') }}"
+  when: (upgrade_categories is defined)
+
+- name: Set dell_dsu_update_all_firmware to True
```

Comment:
Should this be defined before you use it in line 71?

Reply:
No. It was not there before my update. This is the original code: https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/master/tasks/main.yml

No need to init dell_dsu_update_all_firmware to True as by default has the value true. Please check this:
https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/master/defaults/main.yml

Reply:
I did not see that. But then the changes you make even make less sense to me :)

if it defaults to true, why set it to true again `when: (categories_list is defined) and ("all" in categories_list) and (categories_list | length == 1)` you would just not set catagories list if you want to update all firmware? :)

Reply:
I agree. The value of dell_dsu_update_all_firmware is True, so there is no need to set it again true. However, in our ansible playbook, we set dell_dsu_update_all_firmware to False. The idea to set dell_dsu_update_all_firmware when categories_list is set to "all". However, to remove the confusion, I plan to remove this option.

The farmware can be partially update if categories_list is defined with a list. Otherwise, all the framware should be updated if dell_dsu_update_all_firmware is True.

Reply:
well the default case is nothing is overwritten, just defaults defined, in that case it should update by running "dsu --non-interactive --apply-upgrades" I took another look at your code and I don#t think it does that currently. But I only read the code, so I may have missed something

--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L66
not executed because upgrade_categories is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L73
not executed  because upgrade_categories is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L78
not executed because categories_list is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L83
not executed because categories_list is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L89
not executed because categories_list  is undefined
--> https://github.com/CSCfi/ansible-role-dell-firmware-upgrade/blob/386ce3198d04df4902dd88eca0b52c1e3d2c939c/tasks/main.yml#L94
not executed because dell_dsu_arguments is not defined

Reply:
I have simplified the code by don't considering upgrade_categories="all". Now, the code does the following:

1) Update all the firmware if dell_dsu_update_all_firmware=True
2) Otherwise, it updates only some components if upgrade_categories is defined. For instance if upgrade_categories="BIOS,iDRAC", then only these two components should be updated.

Reply:
I think this should work now :) I did not test it, only read the code, so maybe give it a couple of spins with the different scenarios on a real dell server :)

Apart from that: I think you still need to update the readme with the new feature (being able to set upgrade_categories to a string of comma seperated categories that need to match the hash keys in firmware_components) and you should make a note about the breaking change (not being able to overwrite dell_dsu_arguments anymore)

before merging, maybe squash the commits :)

Reply:
I already did the tests with our servers. 

I will update readme and push the changes.

Reply:
I have updated the readme by showing the new features.